### DISCUSSION TITLE
Regarding `IGazeControl::triangulate3DPoint()` and `yarp-device-realsense2`

### DISCUSSION THREAD

**Discussion by paliasgh** [robotology/community#681] [Q&A] [answered]
Hello,

I am using `IGazeControl::triangulate3DPoint()` method for my application. I noticed if the robot's head moves (e.g., rotates to left/right), then the returned 3d point of a fixed object changes significantly. The documentation says the output is in the root reference frame, which does not move in this case, as it is attached to the robot's waist. Can the changes I see be related to distortion that still exists in the camera outputs? In the ideal case, will `triangulate3DPoint()` automatically take care of the head position? 

Also, is there any way to calibrate the returned 3d points other than playing with the eye vergence and adjusting camera calibration parameters? Thank you very much.

**Comment by pattacini** [comment, marked as answer]
Hi @paliasgh 

What is the context of your application?

The documentation[^1] reports in multiple places the following caveat:

> [!note]
The triangulation makes use of the [mid-point method](http://en.wikipedia.org/wiki/Triangulation_(computer_vision)) and thus is deeply affected by uncertainties in the cameras extrinsic parameters and cameras alignment.

Therefore, this method can be applied, essentially, only in simulation[^2] as a tiny error affecting the stereo calibration yields a substantial modification of the triangulated 3D point.

>Can the changes I see be related to distortion that still exists in the camera outputs? 

Not only the distortion but primarily the extrinsic parameters of the stereo rig plus how the rig is (dis-)aligned with the robot's kinematics. This is full of unknowns, eventually, which can be barely estimated/compensated for.

>In the ideal case, will `triangulate3DPoint()` automatically take care of the head position?

Yes. 

>Also, is there any way to calibrate the returned 3d points other than playing with the eye vergence and adjusting camera calibration parameters? Thank you very much.

Abandon all hope... 😄
That's the reason why with the iCub many of us ended up using a RealSense mounted on top of the head.
See https://icub-tech-iit.github.io/documentation/upgrade_kits/realsense_holder/support/.

[^1]: See:
      - https://yarp.it/latest/classyarp_1_1dev_1_1IGazeControl.html#a30cb9848e3d8d92c33bd5a807e1038d9
      - https://robotology.github.io/robotology-documentation/doc/html/group__iKinGazeCtrl.html

[^2]: Examples:
      - https://github.com/vvv-school/assignment_make-it-roll
      - https://github.com/vvv-school/assignment_grasp-it

**Reply by paliasgh** [reply]
Hello @pattacini ,

Thank you very much for your thorough response!

The purpose is to simply locate some objects on a table in front of the robot. I am using ArUco markers and detecting their centers seen from both eyes. I was hoping triangulation works as it gets pixel locations of the exact same spot from each camera. Overall, the values make sense, but the system needs to be heavily calibrated and all the errors you noted play a role.

We think it is time for us to give up using robot cameras and switch to RealSense! We already have a spare D415 model in our lab. I started implementing [yarp-device-realsense2](https://github.com/robotology/yarp-device-realsense2) (this is what we need to get started, right?) while we print the holder. I just have a quick question about that. As part of installation, we should:

> Before proceeding further, please enable the following CMake flags in YARP:
> 
> ENABLE_yarpmod_RGBDSensorWrapper
> ENABLE_yarpmod_RGBDSensorClient
> ENABLE_yarpcar_mjpeg
> ENABLE_yarppm_depthimage_to_rgb

How this should be properly done given that we have `robotology-superbuild` installation? How can we rebuild only YARP with these additions? Thank you very much for all your help.

**Reply by pattacini** [reply]
Hi @paliasgh 

Check this out:
- https://github.com/robotology/robotology-superbuild/tree/master?tab=readme-ov-file#how-do-i-pass-cmake-options-to-the-projects-built-by-the-robotology-superbuild-

That said, it may be convenient to add `yarp-device-realsense2` to the superbuild.
cc @traversaro @Nicogene

**Reply by traversaro** [reply]
> That said, it may be convenient to add `yarp-device-realsense2` to the superbuild.

Yes, I totally agree, see https://github.com/robotology/robotology-superbuild/issues/564 . The issue is kind of stale, but honestly there is really nothing blocking that.

**Reply by paliasgh** [reply]
Hello @traversaro and @pattacini ,

Thanks a lot! I could successfully install `yarp-device-realsense2`. I noticed that in our installation there was no `RGBDSensorWrapper` device. There was a device called `rgbdSensor_nws_yarp`, so I put that one instead in our `RealSense_conf.ini` file to make it work. Maybe documents on `yarp-device-realsense2` need to be updated?

**Reply by paliasgh** [reply]
Hello again,

There is something I need to ask here. When [installing ViSP](https://visp-doc.inria.fr/doxygen/visp-daily/tutorial-install-ubuntu.html) as a requirement of  [realsense-holder-calibration](https://github.com/robotology/realsense-holder-calibration). I receive this yarp-related error:

```
[ 11%] Building CXX object modules/core/CMakeFiles/visp_core.dir/src/display/vpDisplay.cpp.o
In file included from /usr/local/src/robot/visp/modules/core/src/display/vpDisplay.cpp:40:
/usr/local/src/robot/visp/modules/core/include/visp3/core/vpImageConvert.h:57:10: fatal error: yarp/sig/Image.h: No such file or directory
 #include <yarp/sig/Image.h>
          ^~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [modules/core/CMakeFiles/visp_core.dir/build.make:160: modules/core/CMakeFiles/visp_core.dir/src/display/vpDisplay.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:5192: modules/core/CMakeFiles/visp_core.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

Do you have any idea why yarp/sig header can not be imported?

Also, I have some problems/confusions regarding the usage of the `yarp::dev::IRGBDSensor` interface. While realsense device is running (I can see both depth and RGB streams using [simpleRealsense application](https://github.com/robotology/yarp-device-realsense2?tab=readme-ov-file#as-yarpmanager-application)), I can not access anything using `IRGBDSensor` in a code. I have:

```
#include <yarp/dev/IRGBDSensor.h>
IRGBDSensor       *irgbd;
PolyDriver m_rgbdDrv;
```
and then:

```
m_rgbdDrv.view(irgbd);
if (!irgbd)
{   
    yError() << "Error opening iRGBD interface. Device not available";
    return false;
}
```
gives the error. Generally, I am not sure whether I need `yarp::dev::IRGBDSensor` or not. What I am interested in is to get 3D location of a given pixel in the realsense's field of view. Is this something possible through yarp, or do I need to look for an external code for this application?

**Reply by traversaro** [reply]
> Do you have any idea why yarp/sig header can not be imported?

The CMake code of visp related to YARP seems to be quite old (last update 9 years ago) and quite convoluted. In particular, all the logic in https://github.com/lagadic/visp/blob/v3.6.0/modules/core/CMakeLists.txt#L160-L192 seems quite strange, can you try to substitute it with just `list(APPEND opt_libs ${YARP_LIBRARIES})`? Furthermore, are you sure that for https://github.com/robotology/realsense-holder-calibration you need to actually compile visp with USE_YARP enabled?

**Reply by pattacini** [reply]
Regarding `IRGBDSensor`, did you open the driver before working with the I/F through the view?

Here's an example:
- https://github.com/robotology/assistive-rehab/blob/6585d2a0a17e9560e33ad8feb56be2ed4b1bcc43/modules/skeletonRetriever/src/main.cpp#L874-L895

**Reply by paliasgh** [reply]
Hello @traversaro and @pattacini ,

> can you try to substitute it with just list(APPEND opt_libs ${YARP_LIBRARIES})?

That solved the error. Thanks! However, other errors happened when building c++ codes later. I compiled visp without `USE_YARP` with no errors. Let's see if everything works.

> did you open the driver before working with the I/F through the view?

My bad! I did it and it works fine now. Thanks!

I have a question regarding [how to obtain the calibration matrix](https://github.com/robotology/realsense-holder-calibration?tab=readme-ov-file#how-to-obtain-the-calibration-matrix) for our holder. Running `realsense-holder-calibration-process /usr/local/src/robot/visp/visp-build /usr/local/src/robot/realsense_calib 15 8 6 0.02` gives nothing but:

```
Transformation from iCubHeadCenter to RealSense RGB frame:

cat: /usr/local/src/robot/realsense_calib/eMc.txt: No such file or directory
```
Should `eMc.txt` be the output containing the calibration matrix? I also tried making an empty txt file with that name there. With that, it simply gives nothing:

```
Transformation from iCubHeadCenter to RealSense RGB frame:


```


I have already collected data in `/usr/local/src/robot/realsense_calib`:

```
icub@icub26:/usr/local/src/robot/realsense_calib$ ls
camera.xml    image-15.png  image-7.png       pose_fPe_14.yaml  pose_fPe_6.yaml
command.txt   image-1.png   image-8.png       pose_fPe_15.yaml  pose_fPe_7.yaml
image-10.png  image-2.png   image-9.png       pose_fPe_1.yaml   pose_fPe_8.yaml
image-11.png  image-3.png   pose_fPe_10.yaml  pose_fPe_2.yaml   pose_fPe_9.yaml
image-12.png  image-4.png   pose_fPe_11.yaml  pose_fPe_3.yaml
image-13.png  image-5.png   pose_fPe_12.yaml  pose_fPe_4.yaml
image-14.png  image-6.png   pose_fPe_13.yaml  pose_fPe_5.yaml
```
What do you think can be wrong here?

As a note, `Eye-hand_calibration publisher` seems to be working fine with the [given calibration matrices](https://github.com/robotology/realsense-holder-calibration?tab=readme-ov-file#calibration-matrices-of-pre-existing-holders). However, I wanted to obtain our own calibration matrix in case ours is a bit different due to print and installation errors.

**Reply by pattacini** [reply]
Hi @paliasgh 

I've never dealt with this module. Consider debugging it by inserting suitable printouts.
Let me point you to the maintainers:
cc @xEnVrE @gabrielecaddeo

**Reply by xEnVrE** [reply]
Hi @paliasgh 

the script `realsense-holder-calibration-process` that you are running is nothing more than bash script (please find its source code [here](https://github.com/robotology/realsense-holder-calibration/blob/main/src/calibration/scripts/calibrate.sh)). Although not the the best solution, the script is calling the code of two `VISP` tutorials (see lines `23` and `24` for the path to the executables) and clearing out the output, for the sake of clarity, as we are mainly interested in showing the final calibration matrix saved inside `eMc.txt`.

If the file `eMc.txt` does not exist, these are - at least - two possible reasons:
- your `VISP` build was obtained without activating suitable options at the `CMake`-level for building the tutorials - our bad not having it specified in the README - OR;
- the executables are there but the procedure fails for some reason.

In both cases you might want to comment [this](https://github.com/robotology/realsense-holder-calibration/blob/97e4fc3c4242f86810c03e33bc227854be514bd1/src/calibration/scripts/calibrate.sh#L28) line and then `make install` again the `realsense-holder-calibration` repository before re-running the `realsense-holder-calibration-process` script. If you could post the complete output this could help understand a bit.

**Reply by xEnVrE** [reply]
> Furthermore, are you sure that for https://github.com/robotology/realsense-holder-calibration you need to actually compile visp with USE_YARP enabled?

For that, I will update the `README.md` to make clear that indeed for that repository we don't need `USE_YARP` enabled.

**Reply by paliasgh** [reply]
Hello @xEnVrE ,

Thanks for your prompt response! I now see the problem. The `visp` I built from their master branch does not have the required tutorials. [In fact, there is not folder called `calibration` there](https://github.com/lagadic/visp/tree/master/tutorial). The `v3.6.0` branch [seems to have that folder](https://github.com/lagadic/visp/tree/v3.6.0/tutorial/calibration). I will switch now and hopefully the problem will be solved. Thanks again!

**Reply by paliasgh** [reply]
Hello again @xEnVrE ,

Yes, that worked! Thanks again.

There is something to mention: `visp/tutorial/calibration/tutorial-hand-eye-calibration` does not seem to have `--ndata` parameter anymore, [as it is supposed to run using it](https://github.com/robotology/realsense-holder-calibration/blob/f01cbd88f6d0cf631e7db871bd89ef80fd5540bb/src/calibration/scripts/calibrate.sh#L27C1-L28C1). I removed  `--ndata $N_POSES` and could run `realsense-holder-calibration-process` with no issue. 

During the process, `realsense-holder-calibration-process` rejected my data a couple times because collected pictures were not [fully covering the half sphere over the chessboard.](https://visp-doc.inria.fr/doxygen/visp-daily/tutorial-calibration-extrinsic.html#:~:text=Calibrate%20extrinsic-,Step%201%3A%20Acquire%20robot%20poses%2C%20images%20and%20intrinsics,-The%20objective%20here) I changed the location of the cheeseboard so that it finally generated a calibration matrix. With the resulted matrix, however, `realsense-holder-publisher` does not give estimates as good as with [the given matrix](https://github.com/robotology/realsense-holder-calibration?tab=readme-ov-file#calibration-matrices-of-pre-existing-holders). How exactly should I hold the cheeseboard to obtain better results? I have been using calibration poses as in [config_iCubGenova01_holder_tilt.ini](https://github.com/robotology/realsense-holder-calibration/blob/f01cbd88f6d0cf631e7db871bd89ef80fd5540bb/src/calibration/app/conf/config_iCubGenova01_holder_tilt.ini#L12C1-L27C44) and already updated camera intrinsics there. Thanks! Pictures I worked with look like this:

![image-9](https://github.com/robotology/community/assets/59978262/46750940-9c13-45db-91f9-8f931a416996)

**Reply by xEnVrE** [reply]
> With the resulted matrix, however, realsense-holder-publisher does not give estimates as good as with the given matrix. 

Can you elaborate a bit more on this? When do you define the estimates to be "good"? Are you using any additional benchmark to validate the estimates? 

Otherwise, if you are just comparing the output of the publisher, it is normal that you get different estimates than those obtained using the provided matrix as it depends, e.g., on how you mounted the holder, how you collected the pictures, if you are using additional adapters between the holder and the camera, etc.

**Reply by paliasgh** [reply]
Hello @xEnVrE ,

I am using `realsense-holder-publisher` output to convert 3D estimation of a point in real sense frame to robot frame (by a rotation and a translation in the coordinate). In this case, a good estimate is an accurate one (by measuring distances) which is also unaffected by changes in the head position (seeing if head rotations affect results).

Today, I tried putting the chessboard on a table in about 1 meter fart from the robot instead of holding it. I obtained a matrix using which the above process gives almost perfect estimates of 3D location in robot frame! Thanks.

```
0.9984773479  0.03933398265  -0.03867587759  -0.009536235387
-0.008752365042  0.8051971884  0.5929425637  -0.03185936805
0.05446450041  -0.5917012131  0.8043154186  0.0703791129
0  0  0  1
```

**Reply by xEnVrE** [reply]
Glad to know you have managed to achieve the desired performance :)
